jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ChunkAccumulator.java

```diff
@@ -0,0 +1,228 @@
+        {
+            offset += chunk.remaining();
+            BufferUtil.append(buffer.getByteBuffer(), chunk.getByteBuffer());
+            chunk.release();
+        }
+        assert offset == _size;
+        _chunks.clear();
+        _size = 0;
+        return buffer;
+    }
+
+    public void close()
+    {
+        _chunks.forEach(Chunk::release);
+        _chunks.clear();
+        _size = 0;
+    }
+
+    public CompletableFuture<byte[]> readAll(Content.Source source)
+    {
+        return readAll(source, -1);
+    }
+
+    public CompletableFuture<byte[]> readAll(Content.Source source, int maxSize)
+    {
+        CompletableTask<byte[]> task = new AccumulatorTask<>(source, maxSize)
+        {
+            @Override
+            protected byte[] take(ChunkAccumulator accumulator)
+            {
+                return accumulator.take();
+            }
+        };
+        return task.start();
+    }
+
+    /**
+     * @param source The {@link Content.Source} to read
+     * @param pool The {@link ByteBufferPool} to acquire the buffer from, or null for a non {@link Retainable} buffer
+     * @param direct True if the buffer should be direct.
+     * @param maxSize The maximum size to read, or -1 for no limit
+     * @return A {@link CompletableFuture} that will be completed when the complete content is read or
+     * failed if the max size is exceeded or there is a read error.
+     */
+    public CompletableFuture<RetainableByteBuffer> readAll(Content.Source source, ByteBufferPool pool, boolean direct, int maxSize)
+    {
+        CompletableTask<RetainableByteBuffer> task = new AccumulatorTask<>(source, maxSize)
+        {
+            @Override
+            protected RetainableByteBuffer take(ChunkAccumulator accumulator)
+            {
+                return accumulator.take(pool, direct);
+            }
+        };
+        return task.start();
+    }
+
+    private abstract static class AccumulatorTask<T> extends CompletableTask<T>
+    {
+        private final Content.Source _source;
+        private final ChunkAccumulator _accumulator = new ChunkAccumulator();
+        private final int _maxSize;
+
+        private AccumulatorTask(Content.Source source, int maxSize)
+        {
+            _source = source;
+            _maxSize = maxSize;
+        }
+
+        @Override
+        public void run()
+        {
+            try
+            {
+                while (true)
+                {
+                    Chunk chunk = _source.read();
+                    if (chunk == null)
+                    {
+                        _source.demand(this);
+                        return;
+                    }
+
+                    if (Chunk.isFailure(chunk))
+                    {
+                        completeExceptionally(chunk.getFailure());
+                        return;
+                    }
+
+                    if (chunk.hasRemaining())
+                    {
+                        _accumulator.add(chunk);
```

Comment:
There is a check for `isRemaining()` here, and then again inside `add()`.
I think we should get rid of both, so that it'll be easier to specify that the only case you should not call `add()` is when the chunk is a failure.
No need to have a return type for `add()`, and adding EMPTY or EOF chunks has no effect, so that the code is easier for the caller.

Reply:
IAdding EMPTY or EOF does have an effect.  I makes the common case of `ChunkWithContent` followed by `EOF` copy data in `takeRetainableByteBuffer`, when otherwise the optimization for chunks list of 1 would simply return the first chunk.   Now we have to allocate a new buffer and copy the first chunk into it, then notice the next chunk has no content and return the new buffer.
I guess I can make the optimization smarter... but that just means it will be calculating the remaining rather than the add.  Seams like a lot of work to avoid a simple `hasRemaining` call?   Why?

I'll just do the check once inside add.

Reply:
Note that I'm returning boolean from add, so that it follows the common pattern for add methods on collections.  I know we don't use it, but that's what the JIT is for.